File: backend/app/schemas/video.py
"""Video schemas for request/response validation."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


# Request schemas
class PresignedUrlRequest(BaseModel):
    """Request schema for presigned URL generation."""

    filename: str = Field(..., description="Original filename")
    file_size: int = Field(..., gt=0, description="File size in bytes")
    content_type: str = Field(..., description="MIME type of the file")


class PresignedUrlResponse(BaseModel):
    """Response schema for presigned URL generation."""

    presigned_url: str = Field(..., description="Presigned URL for direct upload")
    video_id: UUID = Field(..., description="Video ID created for this upload")
    s3_key: str = Field(..., description="S3 key where the file will be stored")


class VideoCreate(BaseModel):
    """Request schema for creating a video record after upload."""

    video_id: UUID = Field(..., description="Video ID from presigned URL response")
    title: str = Field(..., max_length=255, description="Video title")
    description: str | None = Field(None, description="Video description")
    s3_key: str = Field(..., description="S3 key where the file was uploaded")


class VideoUpdate(BaseModel):
    """Request schema for updating video metadata."""

    title: str | None = Field(None, max_length=255, description="Video title")
    description: str | None = Field(None, description="Video description")


# Response schemas
class VideoRead(BaseModel):
    """Response schema for video data."""

    id: UUID
    user_id: UUID
    title: str
    description: str | None
    file_size: int | None
    format: str | None
    duration: float | None
    resolution: str | None
    s3_key: str | None
    cloudfront_url: str | None
    thumbnail_url: str | None
    status: str
    created_at: datetime
    updated_at: datetime

    class Config:
        """Pydantic config."""

        from_attributes = True

File: frontend/src/types/video.ts
/** Video types. */

import type { VideoStatus } from './video-status'

export interface Video {
  id: string
  user_id: string
  title: string
  description: string | null
  file_size: number | null
  format: string | null
  duration: number | null // Duration in seconds
  resolution: string | null // e.g., "1920x1080"
  s3_key: string | null
  cloudfront_url: string | null
  thumbnail_url: string | null
  status: VideoStatus
  created_at: string
  updated_at: string
}

export interface PresignedUrlRequest {
  filename: string
  file_size: number
  content_type: string
}

export interface PresignedUrlResponse {
  presigned_url: string
  video_id: string
  s3_key: string
}

export interface VideoCreate {
  video_id: string
  title: string
  description?: string | null
  s3_key: string
}

export interface VideoUpdate {
  title?: string
  description?: string | null
}

export interface VideoListParams {
  status?: VideoStatus
  search?: string
  sort_by?: 'created_at' | 'title' | 'duration'
  sort_order?: 'asc' | 'desc'
  limit?: number
  offset?: number
}

export interface DashboardStats {
  total_videos: number
  storage_used_gb: number
  processing_time_minutes: number
  recent_activity_count: number
}

